File: test/extensions/dynamic_modules/test_data/rust/http_stream_callouts_test.rs
use abi::*;
use envoy_proxy_dynamic_modules_rust_sdk::*;

declare_init_functions!(
  init,
  new_http_filter_config_fn,
  new_http_filter_per_route_config_fn
);

fn init() -> bool {
  true
}

fn new_http_filter_config_fn<EC: EnvoyHttpFilterConfig, EHF: EnvoyHttpFilter>(
  _envoy_filter_config: &mut EC,
  name: &str,
  config: &[u8],
) -> Option<Box<dyn HttpFilterConfig<EHF>>> {
  match name {
    "basic_stream_lifecycle" => Some(Box::new(BasicStreamLifecycleConfig {
      cluster_name: String::from_utf8(config.to_owned()).unwrap(),
    })),
    "bidirectional_streaming" => Some(Box::new(BidirectionalStreamingConfig {
      cluster_name: String::from_utf8(config.to_owned()).unwrap(),
    })),
    "multiple_streams" => Some(Box::new(MultipleStreamsConfig {
      cluster_name: String::from_utf8(config.to_owned()).unwrap(),
    })),
    "stream_reset" => Some(Box::new(StreamResetConfig {
      cluster_name: String::from_utf8(config.to_owned()).unwrap(),
    })),
    "upstream_reset" => Some(Box::new(UpstreamResetConfig {
      cluster_name: String::from_utf8(config.to_owned()).unwrap(),
    })),
    _ => panic!("Unknown filter name: {}", name),
  }
}

fn new_http_filter_per_route_config_fn(
  _name: &str,
  _config: &[u8],
) -> Option<Box<dyn std::any::Any>> {
  None
}

// =============================================================================
// Test 1: Basic Stream Lifecycle
// Tests: Start stream, receive headers/data, stream completes successfully.
// =============================================================================

struct BasicStreamLifecycleConfig {
  cluster_name: String,
}

impl<EHF: EnvoyHttpFilter> HttpFilterConfig<EHF> for BasicStreamLifecycleConfig {
  fn new_http_filter(&mut self, _envoy_filter_config: &mut EHF) -> Box<dyn HttpFilter<EHF>> {
    Box::new(BasicStreamLifecycleFilter {
      cluster_name: self.cluster_name.clone(),
      received_response: false,
    })
  }
}

struct BasicStreamLifecycleFilter {
  cluster_name: String,
  received_response: bool,
}

impl<EHF: EnvoyHttpFilter> HttpFilter<EHF> for BasicStreamLifecycleFilter {
  fn on_request_headers(
    &mut self,
    envoy_filter: &mut EHF,
    _end_of_stream: bool,
  ) -> envoy_dynamic_module_type_on_http_filter_request_headers_status {
    // Start an HTTP stream.
    let (result, handle) = envoy_filter.start_http_stream(
      &self.cluster_name,
      vec![
        (":path", b"/test"),
        (":method", b"GET"),
        ("host", b"example.com"),
      ],
      None, // No body.
      true, // end_stream = true.
      5000, // 5 second timeout.
    );

    if result != envoy_dynamic_module_type_http_callout_init_result::Success {
      envoy_filter.send_response(500, vec![("x-error", b"stream_init_failed")], None, None);
      return envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration;
    }

    // Store handle if needed (not storing in this simple test).
    let _ = handle;

    envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration
  }

  fn on_http_stream_headers(
    &mut self,
    _envoy_filter: &mut EHF,
    _stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
    _headers: &[(EnvoyBuffer, EnvoyBuffer)],
    _end_stream: bool,
  ) {
    // Process headers.
  }

  fn on_http_stream_data(
    &mut self,
    _envoy_filter: &mut EHF,
    _stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
    _data: &[EnvoyBuffer],
    _end_stream: bool,
  ) {
    // Process data.
  }

  fn on_http_stream_complete(
    &mut self,
    envoy_filter: &mut EHF,
    _stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
  ) {
    self.received_response = true;
    envoy_filter.send_response(
      200,
      vec![("x-stream", b"success")],
      Some(b"stream_callout_success"),
      None,
    );
  }
}

impl Drop for BasicStreamLifecycleFilter {
  fn drop(&mut self) {
    // Ensure we received the response.
    assert!(
      self.received_response,
      "Stream did not complete successfully"
    );
  }
}

// =============================================================================
// Test 2: Bidirectional Streaming
// Tests: Start stream, send data chunks, receive data chunks, send trailers.
// =============================================================================

struct BidirectionalStreamingConfig {
  cluster_name: String,
}

impl<EHF: EnvoyHttpFilter> HttpFilterConfig<EHF> for BidirectionalStreamingConfig {
  fn new_http_filter(&mut self, _envoy_filter_config: &mut EHF) -> Box<dyn HttpFilter<EHF>> {
    Box::new(BidirectionalStreamingFilter {
      cluster_name: self.cluster_name.clone(),
      stream_handle: std::ptr::null_mut(),
      chunks_received: 0,
    })
  }
}

struct BidirectionalStreamingFilter {
  cluster_name: String,
  stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
  chunks_received: usize,
}

impl<EHF: EnvoyHttpFilter> HttpFilter<EHF> for BidirectionalStreamingFilter {
  fn on_request_headers(
    &mut self,
    envoy_filter: &mut EHF,
    _end_of_stream: bool,
  ) -> envoy_dynamic_module_type_on_http_filter_request_headers_status {
    // Start an HTTP stream with POST method.
    let (result, handle) = envoy_filter.start_http_stream(
      &self.cluster_name,
      vec![
        (":path", b"/stream"),
        (":method", b"POST"),
        ("host", b"example.com"),
        ("content-type", b"application/octet-stream"),
      ],
      None,  // No initial body - we'll stream it.
      false, // end_stream = false.
      10000,
    );

    if result != envoy_dynamic_module_type_http_callout_init_result::Success {
      envoy_filter.send_response(500, vec![("x-error", b"stream_init_failed")], None, None);
      return envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration;
    }

    self.stream_handle = handle;

    // Send chunk 1.
    let chunk1 = b"chunk1";
    let success = unsafe { envoy_filter.send_http_stream_data(handle, chunk1, false) };
    assert!(success);

    // Send chunk 2.
    let chunk2 = b"chunk2";
    let success = unsafe { envoy_filter.send_http_stream_data(handle, chunk2, false) };
    assert!(success);

    // Send trailers to end the stream.
    let trailers = vec![("x-trailer", b"value" as &[u8])];
    let success = unsafe { envoy_filter.send_http_stream_trailers(handle, trailers) };
    assert!(success);

    envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration
  }

  fn on_http_stream_data(
    &mut self,
    _envoy_filter: &mut EHF,
    stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
    _data: &[EnvoyBuffer],
    _end_stream: bool,
  ) {
    assert_eq!(stream_handle, self.stream_handle);
    self.chunks_received += 1;
  }

  fn on_http_stream_complete(
    &mut self,
    envoy_filter: &mut EHF,
    stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
  ) {
    assert_eq!(stream_handle, self.stream_handle);
    let chunks_str = self.chunks_received.to_string();
    envoy_filter.send_response(
      200,
      vec![("x-chunks-received", chunks_str.as_bytes())],
      Some(b"bidirectional_success"),
      None,
    );
  }
}

// =============================================================================
// Test 3: Multiple Streams
// Tests: Start multiple streams concurrently.
// =============================================================================

struct MultipleStreamsConfig {
  cluster_name: String,
}

impl<EHF: EnvoyHttpFilter> HttpFilterConfig<EHF> for MultipleStreamsConfig {
  fn new_http_filter(&mut self, _envoy_filter_config: &mut EHF) -> Box<dyn HttpFilter<EHF>> {
    Box::new(MultipleStreamsFilter {
      cluster_name: self.cluster_name.clone(),
      stream_handles: Vec::new(),
      completed_streams: 0,
    })
  }
}

struct MultipleStreamsFilter {
  cluster_name: String,
  stream_handles: Vec<envoy_dynamic_module_type_http_stream_envoy_ptr>,
  completed_streams: usize,
}

impl<EHF: EnvoyHttpFilter> HttpFilter<EHF> for MultipleStreamsFilter {
  fn on_request_headers(
    &mut self,
    envoy_filter: &mut EHF,
    _end_of_stream: bool,
  ) -> envoy_dynamic_module_type_on_http_filter_request_headers_status {
    // Create 3 concurrent streams.
    for i in 1 ..= 3 {
      let path = format!("/stream{}", i);
      let (result, handle) = envoy_filter.start_http_stream(
        &self.cluster_name,
        vec![
          (":path", path.as_bytes()),
          (":method", b"GET"),
          ("host", b"example.com"),
        ],
        None,
        true, // end_stream = true.
        5000,
      );

      if result == envoy_dynamic_module_type_http_callout_init_result::Success {
        self.stream_handles.push(handle);
      }
    }

    if self.stream_handles.len() != 3 {
      envoy_filter.send_response(500, vec![("x-error", b"stream_init_failed")], None, None);
    }

    envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration
  }

  fn on_http_stream_complete(
    &mut self,
    envoy_filter: &mut EHF,
    stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
  ) {
    assert!(self.stream_handles.contains(&stream_handle));
    self.completed_streams += 1;

    if self.completed_streams == 3 {
      envoy_filter.send_response(200, vec![("x-stream", b"all_success")], None, None);
    }
  }
}

// =============================================================================
// Test 4: Stream Reset
// Tests: Reset an ongoing stream explicitly.
// =============================================================================

struct StreamResetConfig {
  cluster_name: String,
}

impl<EHF: EnvoyHttpFilter> HttpFilterConfig<EHF> for StreamResetConfig {
  fn new_http_filter(&mut self, _envoy_filter_config: &mut EHF) -> Box<dyn HttpFilter<EHF>> {
    Box::new(StreamResetFilter {
      cluster_name: self.cluster_name.clone(),
      stream_handle: std::ptr::null_mut(),
      received_headers: false,
      reset_called: false,
    })
  }
}

struct StreamResetFilter {
  cluster_name: String,
  stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
  received_headers: bool,
  reset_called: bool,
}

impl<EHF: EnvoyHttpFilter> HttpFilter<EHF> for StreamResetFilter {
  fn on_request_headers(
    &mut self,
    envoy_filter: &mut EHF,
    _end_of_stream: bool,
  ) -> envoy_dynamic_module_type_on_http_filter_request_headers_status {
    // Start a stream to a cluster that will be reset.
    let (result, handle) = envoy_filter.start_http_stream(
      &self.cluster_name,
      vec![
        (":path", b"/slow"),
        (":method", b"GET"),
        ("host", b"example.com"),
      ],
      None,
      true, // end_stream = true.
      5000,
    );

    if result != envoy_dynamic_module_type_http_callout_init_result::Success {
      envoy_filter.send_response(500, vec![("x-error", b"stream_init_failed")], None, None);
      return envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration;
    }

    self.stream_handle = handle;
    envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration
  }

  fn on_http_stream_headers(
    &mut self,
    envoy_filter: &mut EHF,
    stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
    _headers: &[(EnvoyBuffer, EnvoyBuffer)],
    _end_stream: bool,
  ) {
    assert_eq!(stream_handle, self.stream_handle);
    self.received_headers = true;

    // Immediately reset the stream after receiving headers.
    unsafe {
      envoy_filter.reset_http_stream(stream_handle);
    }
  }

  fn on_http_stream_reset(
    &mut self,
    envoy_filter: &mut EHF,
    stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
    _reset_reason: envoy_dynamic_module_type_http_stream_reset_reason,
  ) {
    assert_eq!(stream_handle, self.stream_handle);
    self.reset_called = true;

    // Send response indicating reset occurred.
    envoy_filter.send_response(
      200,
      vec![("x-stream", b"reset_ok")],
      Some(b"stream_was_reset"),
      None,
    );
  }
}

impl Drop for StreamResetFilter {
  fn drop(&mut self) {
    assert!(self.received_headers, "Never received headers before reset");
    assert!(self.reset_called, "Reset callback was not called");
  }
}

// =============================================================================
// Test 5: Upstream Reset
// Tests: Start stream, upstream resets connection, receive reset callback.
// =============================================================================

struct UpstreamResetConfig {
  cluster_name: String,
}

impl<EHF: EnvoyHttpFilter> HttpFilterConfig<EHF> for UpstreamResetConfig {
  fn new_http_filter(&mut self, _envoy_filter_config: &mut EHF) -> Box<dyn HttpFilter<EHF>> {
    Box::new(UpstreamResetFilter {
      cluster_name: self.cluster_name.clone(),
      stream_handle: std::ptr::null_mut(),
    })
  }
}

struct UpstreamResetFilter {
  cluster_name: String,
  stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
}

impl<EHF: EnvoyHttpFilter> HttpFilter<EHF> for UpstreamResetFilter {
  fn on_request_headers(
    &mut self,
    envoy_filter: &mut EHF,
    _end_of_stream: bool,
  ) -> envoy_dynamic_module_type_on_http_filter_request_headers_status {
    // Start a stream that we expect to be reset by the upstream.
    let (result, handle) = envoy_filter.start_http_stream(
      &self.cluster_name,
      vec![
        (":path", b"/reset"),
        (":method", b"GET"),
        ("host", b"example.com"),
      ],
      None,
      true,
      5000,
    );

    if result != envoy_dynamic_module_type_http_callout_init_result::Success {
      envoy_filter.send_response(500, vec![("x-error", b"stream_init_failed")], None, None);
      return envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration;
    }

    self.stream_handle = handle;
    envoy_dynamic_module_type_on_http_filter_request_headers_status::StopIteration
  }

  fn on_http_stream_reset(
    &mut self,
    envoy_filter: &mut EHF,
    stream_handle: envoy_dynamic_module_type_http_stream_envoy_ptr,
    _reason: envoy_dynamic_module_type_http_stream_reset_reason,
  ) {
    assert_eq!(stream_handle, self.stream_handle);
    envoy_filter.send_response(
      200,
      vec![("x-reset", b"true")],
      Some(b"upstream_reset"),
      None,
    );
  }
}

File: test/extensions/dynamic_modules/test_data/go/http_stream_callouts_test/http_stream_callouts_test.go
// HTTP stream callouts test module. Mirrors test_data/rust/http_stream_callouts_test.rs.
//
// Exercises the StartHttpStream / SendHttpStreamData / SendHttpStreamTrailers /
// ResetHttpStream API family across five filter shapes:
//
//   basic_stream_lifecycle  - start, receive headers/data, complete
//   bidirectional_streaming - send chunks + trailers, count received chunks
//   multiple_streams        - 3 concurrent streams
//   stream_reset            - reset on first headers callback
//   upstream_reset          - rely on upstream to reset
//
// This module is built but currently has no integration driver — its purpose is to
// exercise the SDK API surface at compile time, paralleling the Rust module of the
// same name.
package main

import (
	"fmt"

	sdk "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go"
	_ "github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/abi"
	"github.com/envoyproxy/envoy/source/extensions/dynamic_modules/sdk/go/shared"
)

func init() {
	sdk.RegisterHttpFilterConfigFactories(map[string]shared.HttpFilterConfigFactory{
		"basic_stream_lifecycle":  &basicStreamLifecycleFactory{},
		"bidirectional_streaming": &bidirectionalStreamingFactory{},
		"multiple_streams":        &multipleStreamsFactory{},
		"stream_reset":            &streamResetFactory{},
		"upstream_reset":          &upstreamResetFactory{},
	})
}

func main() {} //nolint:all

// =============================================================================
// Test 1: Basic Stream Lifecycle
// =============================================================================

type basicStreamLifecycleFactory struct {
	shared.EmptyHttpFilterConfigFactory
}

func (basicStreamLifecycleFactory) Create(_ shared.HttpFilterConfigHandle, c []byte) (shared.HttpFilterFactory, error) {
	return &basicStreamLifecycleFilterFactory{cluster: string(c)}, nil
}

type basicStreamLifecycleFilterFactory struct {
	shared.EmptyHttpFilterFactory
	cluster string
}

func (f *basicStreamLifecycleFilterFactory) Create(h shared.HttpFilterHandle) shared.HttpFilter {
	return &basicStreamLifecycleFilter{handle: h, cluster: f.cluster}
}

type basicStreamLifecycleFilter struct {
	shared.EmptyHttpFilter
	handle           shared.HttpFilterHandle
	cluster          string
	receivedResponse bool
}

func (p *basicStreamLifecycleFilter) OnRequestHeaders(_ shared.HeaderMap, _ bool) shared.HeadersStatus {
	res, _ := p.handle.StartHttpStream(p.cluster,
		[][2]string{{":path", "/test"}, {":method", "GET"}, {"host", "example.com"}},
		nil, true, 5000, p)
	if res != shared.HttpCalloutInitSuccess {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "stream_init_failed"}}, nil, "")
		return shared.HeadersStatusStop
	}
	return shared.HeadersStatusStop
}

func (p *basicStreamLifecycleFilter) OnHttpStreamHeaders(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer, _ bool) {
}
func (p *basicStreamLifecycleFilter) OnHttpStreamData(_ uint64, _ []shared.UnsafeEnvoyBuffer, _ bool) {
}
func (p *basicStreamLifecycleFilter) OnHttpStreamTrailers(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer) {
}
func (p *basicStreamLifecycleFilter) OnHttpStreamComplete(_ uint64) {
	p.receivedResponse = true
	p.handle.SendLocalResponse(200,
		[][2]string{{"x-stream", "success"}},
		[]byte("stream_callout_success"), "")
}
func (p *basicStreamLifecycleFilter) OnHttpStreamReset(_ uint64, _ shared.HttpStreamResetReason) {
}

// =============================================================================
// Test 2: Bidirectional Streaming
// =============================================================================

type bidirectionalStreamingFactory struct {
	shared.EmptyHttpFilterConfigFactory
}

func (bidirectionalStreamingFactory) Create(_ shared.HttpFilterConfigHandle, c []byte) (shared.HttpFilterFactory, error) {
	return &bidirectionalStreamingFilterFactory{cluster: string(c)}, nil
}

type bidirectionalStreamingFilterFactory struct {
	shared.EmptyHttpFilterFactory
	cluster string
}

func (f *bidirectionalStreamingFilterFactory) Create(h shared.HttpFilterHandle) shared.HttpFilter {
	return &bidirectionalStreamingFilter{handle: h, cluster: f.cluster}
}

type bidirectionalStreamingFilter struct {
	shared.EmptyHttpFilter
	handle         shared.HttpFilterHandle
	cluster        string
	streamHandle   uint64
	chunksReceived int
}

func (p *bidirectionalStreamingFilter) OnRequestHeaders(_ shared.HeaderMap, _ bool) shared.HeadersStatus {
	res, id := p.handle.StartHttpStream(p.cluster,
		[][2]string{
			{":path", "/stream"},
			{":method", "POST"},
			{"host", "example.com"},
			{"content-type", "application/octet-stream"},
		},
		nil, false, 10000, p)
	if res != shared.HttpCalloutInitSuccess {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "stream_init_failed"}}, nil, "")
		return shared.HeadersStatusStop
	}
	p.streamHandle = id

	// Send chunks + trailers.
	if !p.handle.SendHttpStreamData(id, []byte("chunk1"), false) {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "send_chunk1"}}, nil, "")
		return shared.HeadersStatusStop
	}
	if !p.handle.SendHttpStreamData(id, []byte("chunk2"), false) {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "send_chunk2"}}, nil, "")
		return shared.HeadersStatusStop
	}
	if !p.handle.SendHttpStreamTrailers(id, [][2]string{{"x-trailer", "value"}}) {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "send_trailers"}}, nil, "")
		return shared.HeadersStatusStop
	}

	return shared.HeadersStatusStop
}

func (p *bidirectionalStreamingFilter) OnHttpStreamHeaders(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer, _ bool) {
}
func (p *bidirectionalStreamingFilter) OnHttpStreamData(id uint64, _ []shared.UnsafeEnvoyBuffer, _ bool) {
	if id == p.streamHandle {
		p.chunksReceived++
	}
}
func (p *bidirectionalStreamingFilter) OnHttpStreamTrailers(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer) {
}
func (p *bidirectionalStreamingFilter) OnHttpStreamComplete(id uint64) {
	if id != p.streamHandle {
		return
	}
	p.handle.SendLocalResponse(200,
		[][2]string{{"x-chunks-received", fmt.Sprintf("%d", p.chunksReceived)}},
		[]byte("bidirectional_success"), "")
}
func (p *bidirectionalStreamingFilter) OnHttpStreamReset(_ uint64, _ shared.HttpStreamResetReason) {
}

// =============================================================================
// Test 3: Multiple Streams
// =============================================================================

type multipleStreamsFactory struct {
	shared.EmptyHttpFilterConfigFactory
}

func (multipleStreamsFactory) Create(_ shared.HttpFilterConfigHandle, c []byte) (shared.HttpFilterFactory, error) {
	return &multipleStreamsFilterFactory{cluster: string(c)}, nil
}

type multipleStreamsFilterFactory struct {
	shared.EmptyHttpFilterFactory
	cluster string
}

func (f *multipleStreamsFilterFactory) Create(h shared.HttpFilterHandle) shared.HttpFilter {
	return &multipleStreamsFilter{handle: h, cluster: f.cluster}
}

type multipleStreamsFilter struct {
	shared.EmptyHttpFilter
	handle           shared.HttpFilterHandle
	cluster          string
	streamHandles    []uint64
	completedStreams int
}

func (p *multipleStreamsFilter) OnRequestHeaders(_ shared.HeaderMap, _ bool) shared.HeadersStatus {
	for i := 1; i <= 3; i++ {
		path := fmt.Sprintf("/stream%d", i)
		res, id := p.handle.StartHttpStream(p.cluster,
			[][2]string{{":path", path}, {":method", "GET"}, {"host", "example.com"}},
			nil, true, 5000, p)
		if res == shared.HttpCalloutInitSuccess {
			p.streamHandles = append(p.streamHandles, id)
		}
	}
	if len(p.streamHandles) != 3 {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "stream_init_failed"}}, nil, "")
	}
	return shared.HeadersStatusStop
}

func (p *multipleStreamsFilter) OnHttpStreamHeaders(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer, _ bool) {
}
func (p *multipleStreamsFilter) OnHttpStreamData(_ uint64, _ []shared.UnsafeEnvoyBuffer, _ bool) {
}
func (p *multipleStreamsFilter) OnHttpStreamTrailers(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer) {
}
func (p *multipleStreamsFilter) OnHttpStreamComplete(id uint64) {
	for _, h := range p.streamHandles {
		if h == id {
			p.completedStreams++
			break
		}
	}
	if p.completedStreams == 3 {
		p.handle.SendLocalResponse(200, [][2]string{{"x-stream", "all_success"}}, nil, "")
	}
}
func (p *multipleStreamsFilter) OnHttpStreamReset(_ uint64, _ shared.HttpStreamResetReason) {}

// =============================================================================
// Test 4: Stream Reset
// =============================================================================

type streamResetFactory struct {
	shared.EmptyHttpFilterConfigFactory
}

func (streamResetFactory) Create(_ shared.HttpFilterConfigHandle, c []byte) (shared.HttpFilterFactory, error) {
	return &streamResetFilterFactory{cluster: string(c)}, nil
}

type streamResetFilterFactory struct {
	shared.EmptyHttpFilterFactory
	cluster string
}

func (f *streamResetFilterFactory) Create(h shared.HttpFilterHandle) shared.HttpFilter {
	return &streamResetFilter{handle: h, cluster: f.cluster}
}

type streamResetFilter struct {
	shared.EmptyHttpFilter
	handle       shared.HttpFilterHandle
	cluster      string
	streamHandle uint64
	receivedHdrs bool
	resetCalled  bool
}

func (p *streamResetFilter) OnRequestHeaders(_ shared.HeaderMap, _ bool) shared.HeadersStatus {
	res, id := p.handle.StartHttpStream(p.cluster,
		[][2]string{{":path", "/slow"}, {":method", "GET"}, {"host", "example.com"}},
		nil, true, 5000, p)
	if res != shared.HttpCalloutInitSuccess {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "stream_init_failed"}}, nil, "")
		return shared.HeadersStatusStop
	}
	p.streamHandle = id
	return shared.HeadersStatusStop
}

func (p *streamResetFilter) OnHttpStreamHeaders(id uint64, _ [][2]shared.UnsafeEnvoyBuffer, _ bool) {
	if id != p.streamHandle {
		return
	}
	p.receivedHdrs = true
	// Immediately reset the stream after receiving headers.
	p.handle.ResetHttpStream(id)
}
func (p *streamResetFilter) OnHttpStreamData(_ uint64, _ []shared.UnsafeEnvoyBuffer, _ bool) {}
func (p *streamResetFilter) OnHttpStreamTrailers(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer)  {}
func (p *streamResetFilter) OnHttpStreamComplete(_ uint64)                                   {}
func (p *streamResetFilter) OnHttpStreamReset(id uint64, _ shared.HttpStreamResetReason) {
	if id != p.streamHandle {
		return
	}
	p.resetCalled = true
	p.handle.SendLocalResponse(200,
		[][2]string{{"x-stream", "reset_ok"}},
		[]byte("stream_was_reset"), "")
}

// =============================================================================
// Test 5: Upstream Reset
// =============================================================================

type upstreamResetFactory struct {
	shared.EmptyHttpFilterConfigFactory
}

func (upstreamResetFactory) Create(_ shared.HttpFilterConfigHandle, c []byte) (shared.HttpFilterFactory, error) {
	return &upstreamResetFilterFactory{cluster: string(c)}, nil
}

type upstreamResetFilterFactory struct {
	shared.EmptyHttpFilterFactory
	cluster string
}

func (f *upstreamResetFilterFactory) Create(h shared.HttpFilterHandle) shared.HttpFilter {
	return &upstreamResetFilter{handle: h, cluster: f.cluster}
}

type upstreamResetFilter struct {
	shared.EmptyHttpFilter
	handle       shared.HttpFilterHandle
	cluster      string
	streamHandle uint64
}

func (p *upstreamResetFilter) OnRequestHeaders(_ shared.HeaderMap, _ bool) shared.HeadersStatus {
	res, id := p.handle.StartHttpStream(p.cluster,
		[][2]string{{":path", "/reset"}, {":method", "GET"}, {"host", "example.com"}},
		nil, true, 5000, p)
	if res != shared.HttpCalloutInitSuccess {
		p.handle.SendLocalResponse(500, [][2]string{{"x-error", "stream_init_failed"}}, nil, "")
		return shared.HeadersStatusStop
	}
	p.streamHandle = id
	return shared.HeadersStatusStop
}

func (p *upstreamResetFilter) OnHttpStreamHeaders(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer, _ bool) {
}
func (p *upstreamResetFilter) OnHttpStreamData(_ uint64, _ []shared.UnsafeEnvoyBuffer, _ bool) {}
func (p *upstreamResetFilter) OnHttpStreamTrailers(_ uint64, _ [][2]shared.UnsafeEnvoyBuffer)  {}
func (p *upstreamResetFilter) OnHttpStreamComplete(_ uint64)                                   {}
func (p *upstreamResetFilter) OnHttpStreamReset(id uint64, _ shared.HttpStreamResetReason) {
	if id != p.streamHandle {
		return
	}
	p.handle.SendLocalResponse(200,
		[][2]string{{"x-reset", "true"}},
		[]byte("upstream_reset"), "")
}
